File: backend/core/api_key_rotation.py
"""Reusable sticky API-key rotation helpers for provider integrations."""

from __future__ import annotations

import logging
from threading import Lock
from dataclasses import dataclass
from typing import Callable, TypeVar


logger = logging.getLogger(__name__)
T = TypeVar("T")


@dataclass(frozen=True)
class ApiKey:
    """A configured API key with its environment-variable slot number."""

    slot: int
    value: str


class StickyApiKeyPool:
    """Try keys from the active slot and rotate only after retryable failures."""

    def __init__(self, provider: str, keys: list[ApiKey]) -> None:
        if not keys:
            raise RuntimeError(f"No valid {provider} API keys are configured.")
        self.provider = provider
        self.keys = tuple(keys)
        self.active_index = 0
        self._lock = Lock()
        logger.info("Loaded %s %s API keys.", len(self.keys), provider)

    def execute(
        self,
        operation: Callable[[str], T],
        *,
        is_retryable: Callable[[Exception], bool],
    ) -> T:
        """Execute an operation against each available key at most once."""

        first_error: Exception | None = None
        with self._lock:
            start_index = self.active_index
        for offset in range(len(self.keys)):
            index = (start_index + offset) % len(self.keys)
            api_key = self.keys[index]
            try:
                result = operation(api_key.value)
                with self._lock:
                    self.active_index = index
                return result
            except Exception as exc:
                if not is_retryable(exc):
                    raise
                first_error = first_error or exc
                logger.warning(
                    "%s quota/rate limit on key #%s.",
                    self.provider,
                    api_key.slot,
                )
                if offset < len(self.keys) - 1:
                    next_index = (start_index + offset + 1) % len(self.keys)
                    next_key = self.keys[next_index]
                    with self._lock:
                        self.active_index = next_index
                    logger.warning(
                        "Switching to %s key #%s.",
                        self.provider,
                        next_key.slot,
                    )

        assert first_error is not None
        raise first_error


def load_api_keys(
    prefix: str,
    *,
    legacy_name: str | None = None,
) -> list[ApiKey]:
    """Load numbered keys, optionally falling back to one legacy variable."""

    import os

    keys: list[ApiKey] = []
    for slot in range(1, 5):
        value = os.getenv(f"{prefix}_{slot}", "").strip()
        if value:
            keys.append(ApiKey(slot=slot, value=value))

    if not keys and legacy_name:
        legacy_value = os.getenv(legacy_name, "").strip()
        if legacy_value:
            keys.append(ApiKey(slot=1, value=legacy_value))
    return keys

File: frontend/src/utils/apiKeyRotation.ts
export type ApiKeyEntry = {
  slot: number;
  value: string;
};

export class StickyApiKeyPool {
  private activeIndex = 0;

  constructor(
    private readonly provider: string,
    private readonly keys: ApiKeyEntry[]
  ) {
    if (keys.length === 0) {
      throw new Error(`No valid ${provider} API keys are configured.`);
    }
    console.info(`Loaded ${keys.length} ${provider} API keys.`);
  }

  async execute<T>(
    operation: (apiKey: string) => Promise<T>,
    isRetryable: (error: unknown) => boolean
  ): Promise<T> {
    let firstError: unknown;
    const startIndex = this.activeIndex;

    for (let offset = 0; offset < this.keys.length; offset += 1) {
      const index = (startIndex + offset) % this.keys.length;
      const apiKey = this.keys[index];
      try {
        const result = await operation(apiKey.value);
        this.activeIndex = index;
        return result;
      } catch (error: unknown) {
        if (!isRetryable(error)) {
          throw error;
        }
        firstError ??= error;
        console.warn(
          `${this.provider} quota/rate limit on key #${apiKey.slot}.`
        );
        if (offset < this.keys.length - 1) {
          const nextIndex = (startIndex + offset + 1) % this.keys.length;
          const nextKey = this.keys[nextIndex];
          this.activeIndex = nextIndex;
          console.warn(`Switching to ${this.provider} key #${nextKey.slot}.`);
        }
      }
    }

    throw firstError;
  }
}

export function loadApiKeys(
  numberedValues: Array<string | undefined>,
  legacyValue?: string
): ApiKeyEntry[] {
  const keys = numberedValues.flatMap((value, index) => {
    const normalized = value?.trim();
    return normalized ? [{ slot: index + 1, value: normalized }] : [];
  });

  const normalizedLegacy = legacyValue?.trim();
  if (keys.length === 0 && normalizedLegacy) {
    return [{ slot: 1, value: normalizedLegacy }];
  }
  return keys;
}
